src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -252,21 +791,43 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                 }
             }
         };
-        debug!("store to var {:?}", index);
-        match &mut self.local_qualif[index] {
-            // this is overly restrictive, because even full assignments do not clear the qualif
-            // While we could special case full assignments, this would be inconsistent with
-            // aggregates where we overwrite all fields via assignments, which would not get
-            // that feature.
-            Some(ref mut qualif) => *qualif = *qualif | self.qualif,
-            // insert new qualification
-            qualif @ None => *qualif = Some(self.qualif),
+
+        let kind = self.mir.local_kind(index);
+        debug!("store to {:?} {:?}", kind, index);
+
+        // Only handle promotable temps in non-const functions.
+        if self.mode == Mode::Fn {
+            if self.mir.local_kind(index) != LocalKind::Temp ||
+               !self.temp_promotion_state[index].is_promotable() {
+                return;
+            }
+        }
+
+        // this is overly restrictive, because even full assignments do not clear the qualif
+        // While we could special case full assignments, this would be inconsistent with
+        // aggregates where we overwrite all fields via assignments, which would not get
+        // that feature.
+        for (per_local, qualif) in &mut self.cx.per_local.as_mut().zip(qualifs).0 {
+            if *qualif {
+                per_local.insert(index);
+            }
+        }
+
+        // Ensure the `IsNotPromotable` qualification is preserved.
+        // NOTE(eddyb) this is actually unnecessary right now, as
+        // we never replace the local's qualif (but we might in
+        // the future) - also, if `IsNotPromotable` only matters
```

Comment:
It does matter in `const fn`. We don't want to end up promoting references to local variables or something weird like that.

Reply:
The "unnecessary right now, as we never replace ..." is the relevant part - vars/temps start out as "not const"/"not promotable", and we current add bits but don't remove them, on assignment.

So right now the code under the comment doesn't need to run, but if someone changes the logic above to also *remove* bits, it would start having an effect.

Maybe I should replace the `insert` with an `assert` that `IsNotPromotable` wasn't removed by the assignment?

Reply:
Yes, that seems like a sensible sanity check.

Reply:
r=me with that assertion